File: python/dynamic-programming/catalan.py
class CatalanNumbers:

    def __init__(self) -> None:
        pass

    def factorial(self,number):
        
        result = 1

        for i in range(number):
            result = result * (i+1)
    
        return result
    
    def catalan(self,n):

        cn = (self.factorial(2*n)) / (self.factorial(n +1) * self.factorial(n))
        return int(cn)
    
    def catalan_list(self,n):
        # init the array
        result_list = [0] * (n + 1)

        # init the base case
        result_list [0]  =  1

        # loop through the figures for the target

        for i in range(n+1):
            cn_i = self.catalan(i)
            result_list[i] = cn_i
        
        return result_list

    
    def catalan_dp(self, number):

        # init the list
        dp = [0] * (number+1)

        # init the base case
        dp[0] = 1

        for i in range(1,number+1):
            #Recurrence relation
            factor = (2*(2*i - 1))/(i+1)
            dp[i] = int(factor * dp[i-1])
        return dp



        
cn = CatalanNumbers()
print(cn.catalan(5))
print(cn.catalan_dp(5))
print(cn.catalan_list(5))

File: javascript/dynamic-programming/catalan.js
const factorialDp = (n) => {
    // init the list
    target_list = [];
    target_list.push(1, 1);
    let result = 1;
    // Only loop if n is larger than 1, else return 1, or -1 if invalid
    if (n >= 2) {
        for (let i = 2; i <= n; i++) {
            let result = i * target_list[i - 1];
            target_list.push(result);
        };
        return target_list.pop();
    } else if (n < 0) {
        return -1;
    } else {
        return 1;
    }

};


const catalanNumbers = (n) => {
    // Catalan Numbers can be worked out as: (2n)! / ((n+1)!(n!))
    return (factorialDp(2 * n)) / ((factorialDp(n + 1)) * (factorialDp(n)))
};



const catalanDp = (n) => {
    // Catalan Numbers satisfy recurrence relations

    // C0 = 1 and Cn = ((2(2n-1))/(n+1)) * cn - 1

    // init the base case

    let arr = [1]

    // Loop for other cases if n > 0

    if (n == 0) {
        return 1;
    } else if (n < 0) {
        return -1;
    } else {
        for (i = 1; i <= n; i++) {
            let factor = ((2 * ((2 * i) - 1)) / (i + 1))
            arr.push(factor * arr[i - 1])

        }

        return arr;

    }
};
console.log(catalanDp(9))